src/transformers/models/zamba2/modular_zamba2.py

```diff
@@ -0,0 +1,1760 @@
+        self.register_buffer("inv_freq", inv_freq, persistent=False)
+
+    @torch.no_grad()
+    def forward(self, x, position_ids):
+        # x: [bs, num_attention_heads, seq_len, head_size]
+        inv_freq_expanded = self.inv_freq[None, :, None].float().expand(position_ids.shape[0], -1, 1)
+        position_ids_expanded = position_ids[:, None, :].float()
+        # Force float32 since bfloat16 loses precision on long contexts
+        # See https://github.com/huggingface/transformers/pull/29285
+        device_type = x.device.type
+        device_type = device_type if isinstance(device_type, str) and device_type != "mps" else "cpu"
+        with torch.autocast(device_type=device_type, enabled=False):
+            freqs = (inv_freq_expanded.float() @ position_ids_expanded.float()).transpose(1, 2)
+            emb = torch.cat((freqs, freqs), dim=-1)
+            cos = emb.cos()
+            sin = emb.sin()
+        return cos.to(dtype=x.dtype), sin.to(dtype=x.dtype)
+
+
+class Zamba2Attention(ZambaAttention):
+    """
+    Multi-headed attention from 'Attention Is All You Need' paper.
+
+    Adapted from transformers.models.mistral.modeling_mistral.MistralAttention:
+    The input dimension here is attention_hidden_size = 2 * hidden_size, and head_dim = attention_hidden_size // num_heads.
+    The extra factor of 2 comes from the input being the concatenation of original_hidden_states with the output of the previous (mamba) layer
+    (see fig. 2 in https://arxiv.org/pdf/2405.16712).
+    Additionally, replaced
+    attn_weights = torch.matmul(query_states, key_states.transpose(2, 3)) / math.sqrt(self.head_dim) with
+    attn_weights = torch.matmul(query_states, key_states.transpose(2, 3)) / math.sqrt(self.head_dim/2)
+    Finally, this attention layer contributes to tied transformer blocks aimed to increasing compute without increasing model size. Because this
+    layer is tied, un-tied LoRA modules are added to the q, k, v projectors to increase expressivity with a small memory overhead.
+    """
+
+    def __init__(
+        self,
+        config: Zamba2Config,
+        layer_idx: Optional[int] = None,
+        num_fwd_mem_blocks: int = None,
+        block_id: int = None,
+    ):
+        super().__init__(config, layer_idx)
+        self.num_fwd_mem_blocks = num_fwd_mem_blocks
+        self.rope_theta = config.rope_theta
+        self.layer_block_map = layer_type_list(config)
+        self.block_id = block_id
+        self.is_causal = True
+
+        if config.use_shared_attention_lora:
+            self.linear_q_lora_A_list = nn.ModuleList([])
+            self.linear_q_lora_B_list = nn.ModuleList([])
+            self.linear_k_lora_A_list = nn.ModuleList([])
+            self.linear_k_lora_B_list = nn.ModuleList([])
+            self.linear_v_lora_A_list = nn.ModuleList([])
+            self.linear_v_lora_B_list = nn.ModuleList([])
```

Comment:
We'd be happy to update this if strictly necessary, although this would also involve converting all the checkpoints on the hub. The use of `nn.Sequential` will inherently change the weight keys as they currently reflect the separation into `A` and `B` layers. For example: `linear_q_lora_A_list.<index>.weight` would change to something like `linear_q_lora_list.<index>.<index>.weight`. Please let us know if there is still a need for this to be modified.

Reply:
done as you suggested [in this commit](https://github.com/huggingface/transformers/pull/34517/commits/f1b31a13289255e55f612041dea9df4a467831e6#diff-94297873668a42780c28a94f5b612feef4b51307c11decbb9794f79a2b086f7a)